lib/src/rules/weak_mode_unnecessary_null_checks.dart

```diff
@@ -0,0 +1,132 @@
+
+class WeakModeUnnecessaryNullChecks extends LintRule implements NodeLintRule {
+  WeakModeUnnecessaryNullChecks()
+      : super(
+            name: 'weak_mode_unnecessary_null_checks',
+            description: _desc,
+            details: _details,
+            group: Group.style);
+
+  @override
+  void registerNodeProcessors(
+      NodeLintRegistry registry, LinterContext context) {
+    if (context.isEnabled(Feature.non_nullable)) {
+      return;
+    }
+
+    final visitor = _Visitor(this, context);
+    registry.addPropertyAccess(this, visitor);
+    registry.addMethodInvocation(this, visitor);
+    registry.addPostfixExpression(this, visitor);
+    registry.addBinaryExpression(this, visitor);
+  }
+}
+
+class _Visitor extends SimpleAstVisitor<void> {
+  _Visitor(this.rule, this.context);
+
+  final LintRule rule;
+  final LinterContext context;
+
+  @override
+  void visitPropertyAccess(PropertyAccess node) {
+    if (node.operator.type == TokenType.QUESTION_PERIOD &&
+        isNonNullable(node.target)) {
+      rule.reportLintForToken(node.operator);
+    }
+  }
+
+  @override
+  void visitMethodInvocation(MethodInvocation node) {
+    if (node.operator?.type == TokenType.QUESTION_PERIOD &&
+        isNonNullable(node.target)) {
+      rule.reportLintForToken(node.operator);
+    }
+  }
+
+  @override
+  void visitPostfixExpression(PostfixExpression node) {
+    if (node.operator.type == TokenType.QUESTION &&
+        isNonNullable(node.operand)) {
+      rule.reportLintForToken(node.operator);
+    }
+  }
+
+  @override
+  void visitBinaryExpression(BinaryExpression node) {
+    final operands = [
+      node.leftOperand,
+      node.rightOperand,
+    ].map((e) => e.unParenthesized).toList();
+    if ((node.operator.type == TokenType.EQ_EQ ||
+            node.operator.type == TokenType.BANG_EQ) &&
+        operands.where((e) => e.staticType.isDartCoreNull).length == 1) {
+      final operand = operands.firstWhere((e) => !e.staticType.isDartCoreNull);
+      if (isNonNullable(operand)) {
+        rule.reportLint(node);
+      }
+    }
+  }
+
+  bool isNonNullable(Expression expression) {
+    if (expression.staticType.isDynamic) {
+      return false;
+    }
+    Element element;
+    if (expression is SimpleIdentifier) {
+      element = expression.staticElement;
+    } else if (expression is MethodInvocation) {
+      element = expression.methodName.staticElement;
+    } else if (expression is PropertyAccess) {
+      element = expression.propertyName.staticElement;
+    }
+    if (element == null ||
+        !element.library.featureSet.isEnabled(Feature.non_nullable)) {
+      return false;
+    }
+    final node = getNode(element);
+    TypeAnnotation type;
+    if (node is FunctionDeclaration) type = node.returnType;
+    if (node is MethodDeclaration) type = node.returnType;
+    if (element is PropertyAccessorElement && element.isSynthetic) {
+      type = (getNode(element.variable).parent as VariableDeclarationList).type;
```

Comment:
The method `getNode` is fairly expensive, so consider instead doing something like
```
if (element is PropertyAccessorElement && element.isSynthetic) {
  element = element.variable;
}
```
before line 117 and replacing this branch with something like
```
if (node is FieldDeclaration) ...
```

Also consider using an `else if` structure here to make it clear that only one of the conditions is expected to be true.